### PR TITLE
Add code action for RootElementTypeMustMatchDoctypedecl

### LINKED ISSUE #561: Code actions for RootElementTypeMustMatchDoctypedecl
As mentioned in #537 

Given this XML file:

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE efgh [ 
<!ELEMENT abcd (#PCDATA)>
<!ELEMENT efgh (#PCDATA)>
]> 
<abcd/>
```
A CodeAction could be replace `abcd` with `efgh`. This CodeAction should take care of the 2 cases: 
* `<abcd/>` should be replaced with `<efgh/>`
* `<abcd></abcd>` should be replaced with `<efgh></efgh>`

### PR DESCRIPTION
**Pull request by xorye**
Fixes #561 

Test with:
```
<?xml version="1.0" encoding="UTF-8" ?>
<!DOCTYPE abcd [
  <!ELEMENT abcd (#PCDATA)>
  <!ELEMENT efgh (#PCDATA)>
]>
<efgh>
```

image

image

Signed-off-by: David Kwon <dakwon@redhat.com>

### CODE REVIEW

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/RootElementTypeMustMatchDoctypedeclCodeAction.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+import org.eclipse.lemminx.utils.XMLPositionUtility;
+import org.eclipse.lsp4j.CodeAction;
+import org.eclipse.lsp4j.Diagnostic;
+import org.eclipse.lsp4j.Range;
+
+/**
+ * Code action for RootElementTypeMustMatchDoctypedecl
+ * 
+ * This class depends on the diagnostic message (Diagnostic#message) to 
+ * determine the correct root name.
```
Comment (xorye) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/RootElementTypeMustMatchDoctypedeclCodeAction.java:32]: In this class, I retrieve the correct root name from the error message string, which is not robust.

Maybe I should infer the correct root name from the DOMDocument?

Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/RootElementTypeMustMatchDoctypedeclCodeAction.java:32]: I agree it's not robust, but you must do that for `getCurrentRoot` and `getDoctypeRoot`, so I would say "keep your code like this". LSP should be improved to add a data object in the diagnostic like  we have with CodeLens#data

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/RootElementTypeMustMatchDoctypedeclCodeAction.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+		}
+		int postMatch = message.indexOf(postText);
+		if (postMatch < 0) {
+			return null;
+		}
+
+		return message.substring(preMatch + preText.length(), postMatch);
+	}
+
+	private String getDoctypeRoot(String message) {
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/RootElementTypeMustMatchDoctypedeclCodeAction.java:95]: Add javadoc + set this method as static

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/RootElementTypeMustMatchDoctypedeclCodeAction.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+		replaceRanges.add(XMLPositionUtility.selectStartTagName(root));
+		replaceTexts.add(newText);
+
+		if (root.isClosed() && !root.isSelfClosed())  {
+			replaceRanges.add(XMLPositionUtility.selectEndTagName(root));
+			replaceTexts.add(newText);
+		}
+	}
+
+	private String getCurrentRoot(String message) {
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/RootElementTypeMustMatchDoctypedeclCodeAction.java:80]: Add javadoc + set this method as static

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/commons/CodeActionFactory.java [hunk 4]
```diff
@@ -76,16 +77,31 @@ public static CodeAction insert(String title, Position position, String insertTe
 
 	public static CodeAction replace(String title, Range range, String replaceText, TextDocumentItem document,
 			Diagnostic diagnostic) {
+		return replace(title, Collections.singletonList(range), Collections.singletonList(replaceText), document,
+				diagnostic);
+	}
+
+	public static CodeAction replace(String title, List<Range> range, List<String> replaceText, TextDocumentItem document,
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/commons/CodeActionFactory.java:84]: I think you should have this signature:

```
public static CodeAction replace(String title, List<TextEdit> replace, TextDocumentItem document,
			Diagnostic diagnostic) {
```

**Comment by angelozerr** [comment]
Works like a charm, great job @xorye !

**Pull request merged**